src/lib/codec/hex/hex.cpp

```diff
@@ -53,27 +60,36 @@ std::string hex_encode(const uint8_t input[], size_t input_length, bool uppercas
 namespace {
 
 uint8_t hex_char_to_bin(char input) {
-   const uint8_t c = static_cast<uint8_t>(input);
+   // Assumes each byte is either 0x00 or 0x80
+   auto index_of_first_set_byte = [](uint64_t v) {
+      return ((((v - 1) & 0x0101010101010101) * 0x0101010101010101) >> 56) - 1;
+   };
+
+   auto swar_in_range = [](uint64_t v, uint64_t lower, uint64_t upper) -> uint64_t {
+      // This assumes that the high bits of each byte of upper/lower are unset
+      constexpr uint64_t hi = 0x8080808080808080;
+      constexpr uint64_t lo7 = 0x7F7F7F7F7F7F7F7F;
 
-   const auto is_alpha_upper = CT::Mask<uint8_t>::is_within_range(c, uint8_t('A'), uint8_t('F'));
-   const auto is_alpha_lower = CT::Mask<uint8_t>::is_within_range(c, uint8_t('a'), uint8_t('f'));
-   const auto is_decimal = CT::Mask<uint8_t>::is_within_range(c, uint8_t('0'), uint8_t('9'));
+      const uint64_t sub = ((v | hi) - (lower & lo7)) ^ ((v ^ (~lower)) & hi);
+      const uint64_t a_lo = sub & lo7;
+      const uint64_t a_hi = sub & hi;
+      return (lo7 - a_lo + upper) & hi & ~a_hi;
+   };
 
-   const auto is_whitespace =
-      CT::Mask<uint8_t>::is_any_of(c, {uint8_t(' '), uint8_t('\t'), uint8_t('\n'), uint8_t('\r')});
+   // Starting value: '0', 'a', 'A', ' ', '\n', '\t', '\r'
+   constexpr uint64_t v_lo = 0x00306141200a090d;
+   // Valid ranges: 10, 6, 6, 1, 1, 1, 1
+   constexpr uint64_t v_range = 0x0A060601010101;
```

Comment:
Similar to the base64 version: I find this resolves the "magic numbers" quite nicely.

```suggestion
   // Starts of valid value ranges (v_lo) and their lengths (v_range)
   constexpr uint64_t v_lo = make_uint64(0, '0', 'a', 'A', ' ', '\n', '\t', '\r');
   constexpr uint64_t v_range = make_uint64(0, 10, 6, 6, 1, 1, 1, 1);
```